File: python/leetcode/algorithms/dp/matrix.py
"""Good DP problems
"""

import heapq
import random
import math
import bisect
from typing import *
from math import inf, factorial, gcd, lcm
from functools import lru_cache, cache
from heapq import heapify, heappush, heappop
from itertools import accumulate, permutations, combinations
from collections import Counter, deque, defaultdict, OrderedDict
from sortedcontainers import SortedSet, SortedList, SortedDict


class Solution:

    # 3459. Length of Longest V-Shaped Diagonal Segment
    def lenOfVDiagonal(self, grid: List[List[int]]) -> int:
        R = len(grid)
        C = len(grid[0])
        directions = {
            (-1, 1): (1, 1),
            (1, -1): (-1, -1),
            (1, 1): (1, -1),
            (-1, -1): (-1, 1),
        }

        @cache
        def dp(r, c, turned, dir):
            res = 0
            cur = grid[r][c]
            nxt = 2
            if cur == 2:
                nxt = 0
            nr = dir[0] + r
            nc = dir[1] + c
            if 0 <= nr < R and 0 <= nc < C:
                if nxt == grid[nr][nc]:
                    res = dp(nr, nc, turned, dir)
            if not turned:
                dirs = directions[dir]
                nr = dirs[0] + r
                nc = dirs[1] + c
                if 0 <= nr < R and 0 <= nc < C:
                    if nxt == grid[nr][nc]:
                        res = max(dp(nr, nc, True, dirs), res)

            return res + 1

        res = 0
        for r in range(R):
            for c in range(C):
                if grid[r][c] == 1:
                    for k in directions.keys():
                        res = max(res, dp(r, c, False, k))
        return res

    # 931 Minimum Falling Path Sum
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        N = len(matrix)
        directions = ((1, 0), (1, 1), (1, -1))

        @cache
        def dp(r, c):
            if r == N - 1:
                return matrix[r][c]
            res = inf
            for dr, dc in directions:
                nr = dr + r
                nc = dc + c
                if 0 <= nr < N and 0 <= nc < N:
                    res = min(res, dp(nr, nc))
            return res + matrix[r][c]

        best = inf
        for c in range(N):
            best = min(dp(0, c), best)
        return best

File: java/dsa/src/main/java/com/dsa/leetcode/dp/Matrix.java
package com.dsa.leetcode.dp;

import java.util.HashMap;
import java.util.Map;

@SuppressWarnings("all")
public class Matrix {

    int MAX_VAL = Integer.MAX_VALUE;

    // 931. Minimum Falling Path Sum
    public class P931 {

        int N;
        Map<String, Integer> memo = new HashMap<>();
        int[][] directions = { { 1, 0 }, { 1, 1 }, { 1, -1 } };

        public int minFallingPathSum(int[][] matrix) {
            this.N = matrix.length;
            int res = MAX_VAL;
            for (int i = 0; i < this.N; i++) {
                res = Math.min(res, dp(0, i, matrix));
            }
            return res;
        }

        int dp(int r, int c, int[][] matrix) {
            if (memo.containsKey(r + "," + c)) {
                return memo.get(r + "," + c);
            }
            int val = matrix[r][c];
            if (r == this.N - 1) {
                return val;
            }
            int res = MAX_VAL;
            for (int i = 0; i < 3; i++) {
                int nr = this.directions[i][0] + r;
                int nc = this.directions[i][1] + c;
                if (0 <= nr && nr < this.N && 0 <= nc && nc < this.N) {
                    int out = dp(nr, nc, matrix);
                    res = Math.min(out, res);
                }
            }
            memo.put(r + "," + c, val + res);
            return val + res;
        }
    }

    // 3459. Length of Longest V-Shaped Diagonal Segment
    public class P3459 {

        int R;
        int C;
        int[][] grid;
        static Integer[][][][][] memo;
        static final Map<Integer, int[]> dirs = new HashMap<>();
        static {
            dirs.put(-99, new int[] { 1, 1 });
            dirs.put(99, new int[] { -1, -1 });
            dirs.put(101, new int[] { 1, -1 });
            dirs.put(-101, new int[] { -1, 1 });
        }

        public int lenOfVDiagonal(int[][] grid1) {
            R = grid1.length;
            C = grid1[0].length;
            memo = new Integer[R][C][2][3][3];
            grid = grid1;
            int res = 0;
            for (int r = 0; r < R; r++) {
                for (int c = 0; c < C; c++) {
                    if (grid[r][c] == 1) {
                        for (int[] dir : dirs.values()) {
                            res = Math.max(res, dp(r, c, 0, dir));
                        }
                    }
                }
            }
            return res;
        }

        int dp(int r, int c, int turned, int[] dir) {
            if (memo[r][c][turned][dir[0] + 1][dir[1] + 1] != null) {
                return memo[r][c][turned][dir[0] + 1][dir[1] + 1];
            }
            int nxt = grid[r][c] == 2 ? 0 : 2;
            int nr = r + dir[0];
            int nc = c + dir[1];
            int res = 0;
            if (0 <= nr && nr < R && 0 <= nc && nc < C && nxt == grid[nr][nc]) {
                res = dp(nr, nc, turned, dir);
            }
            if (turned == 0) {
                int[] ndir = dirs.get(dir[0] * 100 + dir[1]);
                nr = r + ndir[0];
                nc = c + ndir[1];
                if (0 <= nr && nr < R && 0 <= nc && nc < C && nxt == grid[nr][nc]) {
                    res = Math.max(res, dp(nr, nc, 1, ndir));
                }
            }
            memo[r][c][turned][dir[0] + 1][dir[1] + 1] = res + 1;
            return res + 1;
        }
    }

}
